include/ensmallen_bits/ada_bound/ada_bound_update.hpp

```diff
@@ -0,0 +1,207 @@
+  double& Epsilon() { return epsilon; }
+
+  //! Get the smoothing parameter.
+  double Beta1() const { return beta1; }
+  //! Modify the smoothing parameter.
+  double& Beta1() { return beta1; }
+
+  //! Get the second moment coefficient.
+  double Beta2() const { return beta2; }
+  //! Modify the second moment coefficient.
+  double& Beta2() { return beta2; }
+
+  //! Get the current iteration number.
+  size_t Iteration() const { return iteration; }
+  //! Modify the current iteration number.
+  size_t& Iteration() { return iteration; }
+
+  /**
+   * The UpdatePolicyType policy classes must contain an internal 'Policy'
+   * template class with two template arguments: MatType and GradType.  This is
+   * instantiated at the start of the optimization, and holds parameters
+   * specific to an individual optimization.
+   */
+  template<typename MatType, typename GradType>
+  class Policy
+  {
+   public:
+    /**
+     * This constructor is called by the SGD Optimize() method before the start
+     * of the iteration update process.
+     *
+     * @param parent AdaBoundUpdate object.
+     * @param rows Number of rows in the gradient matrix.
+     * @param cols Number of columns in the gradient matrix.
+     */
+    Policy(AdaBoundUpdate& parent, const size_t rows, const size_t cols) :
+        parent(parent), first(true), initialStepSize(0)
+    {
+      m.zeros(rows, cols);
+      v.zeros(rows, cols);
+    }
+
+    /**
+     * Update step for AdaBound.
+     *
+     * @param iterate Parameters that minimize the function.
+     * @param stepSize Step size to be used for the given iteration.
+     * @param gradient The gradient matrix.
+     */
+    void Update(MatType& iterate,
+                const double stepSize,
+                const GradType& gradient)
+    {
+      // Convenience typedefs.
+      typedef typename MatType::elem_type ElemType;
+
+      // Save the initial step size.
+      if (first)
+      {
+        first = false;
+        initialStepSize = stepSize;
+      }
+
+      // Increment the iteration counter variable.
+      ++parent.iteration;
+
+      // Decay the first and second moment running average coefficient.
+      m *= parent.beta1;
+      m += (1 - parent.beta1) * gradient;
+
+      v *= parent.beta2;
+      v += (1 - parent.beta2) * (gradient % gradient);
+
+      const ElemType biasCorrection1 = 1.0 - std::pow(parent.beta1,
+          parent.iteration);
+      const ElemType biasCorrection2 = 1.0 - std::pow(parent.beta2,
+          parent.iteration);
+
+      const ElemType fl = parent.finalLr * stepSize / initialStepSize;
+      const ElemType lower = fl * (1.0 - 1.0 / (parent.gamma *
+          parent.iteration + 1));
+      const ElemType upper = fl * (1.0 + 1.0 / (parent.gamma *
+          parent.iteration));
+
+       // Applies bounds on actual learning rate.
+      iterate -= arma::clamp((stepSize *
+          std::sqrt(biasCorrection2) / biasCorrection1) / (arma::sqrt(v) +
```

Comment:
The paper is rather confusing here.  Their notation is `V_t = diag(v_t)` but I truly don't understand the need for the `diag()` as `v_t` is a vector.  Anyway if I just forget about the `diag()` or assume it is the identity, which I think is the right thing to do, this implementation seems correct.

Reply:
`v_t` could be a matrix if the `gradient` is a matrix as well, so I will add the `diag`; it shouldn't matter.

Reply:
I'm not sure that's right though.  In the paper they restrict the parameters to be a vector, and the intention of Adam is to scale the momentum of each individual parameter.  It seems like then every element should be used, not just the diagonal.

Reply:
https://openreview.net/forum?id=Bkg3g2R9FX&noteId=S1eizWWuCQ

But you are right, it doesn't really make sense. Also I can't see they use it in there reference implementation.

Reply:
Okay, I think it is a poor and confusing response that the authors gave, but I think it gave me enough of a hint to understand.  In their case `diag()` does not mean "take the diagonal vector of a matrix" like in Armadillo but instead it must be more like Armadillo's `diagmat()` where the meaning is "create a matrix with this vector on the diagonal".

Then, the concept of `sqrt()` is equivalent to element-wise square root: if `A` is diagonal, then `sqrt(A)_{ii} == sqrt(A_ii)`.

But then they must be implicitly intending that the expression `(\alpha / \sqrt(V_t))` in line 6 of the algorithm has an implicit flattening of the 2-d result: i.e., `\hat{eta}_t = Clip(vectorise(\alpha / \sqrt(V_t)), \eta_l(t), \eta_u(t))` (to abuse notation a little bit).

So, in the end, I think this means your original implementation was correct... let me know if you think otherwise.